single_kernel_mongo/managers/config.py

```diff
@@ -0,0 +1,338 @@
+#!/usr/bin/env python3
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Manager for handling Mongo configuration."""
+
+import logging
+import time
+from abc import ABC, abstractmethod
+from itertools import chain
+
+from ops import Container
+from typing_extensions import override
+
+from single_kernel_mongo.config.literals import (
+    LOCALHOST,
+    MongoPorts,
+    RoleEnum,
+    Substrates,
+)
+from single_kernel_mongo.config.models import AuditLogConfig, LogRotateConfig, Role
+from single_kernel_mongo.core.structured_config import MongoConfigModel, MongoDBRoles
+from single_kernel_mongo.core.workload import WorkloadBase
+from single_kernel_mongo.exceptions import WorkloadServiceError
+from single_kernel_mongo.state.charm_state import CharmState
+from single_kernel_mongo.utils.mongodb_users import BackupUser, MonitorUser
+from single_kernel_mongo.workload import (
+    get_logrotate_workload_for_substrate,
+    get_mongodb_exporter_workload_for_substrate,
+    get_pbm_workload_for_substrate,
+)
+from single_kernel_mongo.workload.log_rotate_workload import LogRotateWorkload
+
+logger = logging.getLogger(__name__)
+
+
+class CommonConfigManager(ABC):
+    """A generic config manager for a workload."""
+
+    config: MongoConfigModel
+    workload: WorkloadBase
+    state: CharmState
+
+    def set_environment(self):
+        """Write all parameters in the environment variable."""
+        if self.workload.env_var != "":
+            parameters = chain.from_iterable(self.build_parameters())
+            self.workload.update_env(parameters)
+
+    def get_environment(self) -> str:
+        """Gets the environment for the defined service."""
+        env = self.workload.get_env()
+        return env[self.workload.env_var]
+
+    @abstractmethod
+    def build_parameters(self) -> list[list[str]]:  # pragma: nocover
+        """Builds the parameters list."""
+        ...
+
+
+class BackupConfigManager(CommonConfigManager):
+    """Config manager for PBM."""
+
+    def __init__(
+        self,
+        substrate: Substrates,
+        role: Role,
+        config: MongoConfigModel,
+        state: CharmState,
+        container: Container | None,
```

Comment:
noob question what benefit of creating it only once?

Reply:
Slower memory footprint, and if ever there were some local updates of a variable in the Container, passing it allows us to always have the "correct" version of it.
It should not happen here because a `Container` is just a boilerplate to an HTTP API, but I'd rather be sure :)

Reply:
Thanks so much for this explanation and for your expert knowledge here! Your expert knowledge helps us improve the charm performance! 

I think it would be beneficial to add a comment explaining that. Or even better. I short doc with 1-5 sentences in the repo. This way the knowledge persists